components/eventing-controller/reconciler/subscription/reconciler.go

```diff
@@ -0,0 +1,965 @@
+		return false, nil
+	}
+
+	// update the status
+	subscription.Status.Conditions = desiredConditions
+	subscription.Status.Ready = isReady
+	return true, nil
+}
+
+// emitConditionEvent emits a kubernetes event and sets the event type based on the Condition status
+func (r *Reconciler) emitConditionEvent(subscription *eventingv1alpha1.Subscription, condition eventingv1alpha1.Condition) {
+	eventType := corev1.EventTypeNormal
+	if condition.Status == corev1.ConditionFalse {
+		eventType = corev1.EventTypeWarning
+	}
+	r.recorder.Event(subscription, eventType, string(condition.Reason), condition.Message)
+}
+
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&eventingv1alpha1.Subscription{}).
+		Watches(&source.Kind{Type: &apigatewayv1alpha1.APIRule{}}, r.getAPIRuleEventHandler()).
+		Complete(r)
+}
+
+// getAPIRuleEventHandler returns an APIRule event handler.
+func (r *Reconciler) getAPIRuleEventHandler() handler.EventHandler {
+	eventHandler := func(eventType, name, namespace string, q workqueue.RateLimitingInterface) {
+		log := r.Log.WithValues("event", eventType, "kind", "APIRule", "name", name, "namespace", namespace)
+		if err := r.handleAPIRuleEvent(name, namespace, q, log); err != nil {
+			log.Error(err, "Failed to handle APIRule Event, requeue event again")
+			q.Add(reconcile.Request{NamespacedName: k8stypes.NamespacedName{Name: name, Namespace: namespace}})
+		}
+	}
+
+	return handler.Funcs{
+		CreateFunc: func(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+			eventType, name, namespace := "Create", e.Meta.GetName(), e.Meta.GetNamespace()
+			eventHandler(eventType, name, namespace, q)
+		},
+		UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+			eventType, name, namespace := "Update", e.MetaNew.GetName(), e.MetaNew.GetNamespace()
+			eventHandler(eventType, name, namespace, q)
+		},
+		DeleteFunc: func(e event.DeleteEvent, q workqueue.RateLimitingInterface) {
+			eventType, name, namespace := "Delete", e.Meta.GetName(), e.Meta.GetNamespace()
+			eventHandler(eventType, name, namespace, q)
+		},
+		GenericFunc: func(e event.GenericEvent, q workqueue.RateLimitingInterface) {
+			eventType, name, namespace := "Generic", e.Meta.GetName(), e.Meta.GetNamespace()
+			eventHandler(eventType, name, namespace, q)
+		},
+	}
+}
+
+// handleAPIRuleEvent handles APIRule event.
+func (r *Reconciler) handleAPIRuleEvent(name, namespace string, q workqueue.RateLimitingInterface, log logr.Logger) error {
+	// skip not relevant APIRules
+	if !isRelevantAPIRuleName(name) {
+		return nil
+	}
+
+	log.Info("Handle APIRule Event")
+
+	// try to get the APIRule from the API server
+	ctx := context.Background()
+	apiRule := &apigatewayv1alpha1.APIRule{}
+	key := k8stypes.NamespacedName{Name: name, Namespace: namespace}
+	if err := r.Client.Get(ctx, key, apiRule); err != nil {
+		log.Info("Cannot get APIRule")
```

Comment:
Shouldn't we just silently ignore it if the APIRule was not found ? Also I think we need to return an error if there was a problem (other than not found). What do you think?

Reply:
Done.